From 858e6c9fbb51c1a408c2900e4e5be82554eb78c5 Mon Sep 17 00:00:00 2001
From: Nick Avramoussis <nna@dneg.com>
Date: Tue, 26 Mar 2019 11:29:43 +0000
Subject: [PATCH] Reworked cmake boost_python handling such that versioned
 suffixed libraries are prioritized, falling back to the non-version suffixed
 library if available

Signed-off-by: Nick Avramoussis <nna@dneg.com>
---
 CMakeLists.txt                |  3 ---
 openvdb/python/CMakeLists.txt | 32 ++++++++++++++++++++++++++------
 2 files changed, 26 insertions(+), 9 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 15656cd5b..f57e5ec30 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -316,9 +316,6 @@ OPTION ( ILMBASE_NAMESPACE_VERSIONING "Set the expected names of ILM Base librar
 CMAKE_DEPENDENT_OPTION ( OPENVDB_DISABLE_BOOST_IMPLICIT_LINKING
   "Disable the implicit linking of Boost libraries on Windows" ON "WIN32" OFF )
 OPTION ( USE_SYSTEM_LIBRARY_PATHS "Build with system library paths" ON )
-CMAKE_DEPENDENT_OPTION ( BOOST_PYTHON_NAMESPACE_VERSIONING
-  "Set the expected names of Boost.Python libraries to be version suffixed"
-  ON "OPENVDB_BUILD_PYTHON_MODULE" ON )
 
 SET ( OPENVDB_ABI_VERSION_NUMBER "" CACHE STRING "Build for compatibility with version N of the
   OpenVDB Grid ABI, where N is 3, 4, 5 etc. (some newer features will be disabled)." )
diff --git a/openvdb/python/CMakeLists.txt b/openvdb/python/CMakeLists.txt
index c74a80127..68c99ca94 100644
--- a/openvdb/python/CMakeLists.txt
+++ b/openvdb/python/CMakeLists.txt
@@ -113,16 +113,36 @@ ELSE ()
   LIST ( APPEND OPENVDB_PYTHON_DEPS Python::Python )
 ENDIF ()
 
-IF ( BOOST_PYTHON_NAMESPACE_VERSIONING )
-  SET ( BOOST_PYTHON_LIB "python${Python_VERSION_MAJOR}${Python_VERSION_MINOR}" )
-ELSE ()
-  SET ( BOOST_PYTHON_LIB "python" )
-ENDIF ()
+# Boost
+
+FIND_PACKAGE ( Boost ${MINIMUM_BOOST_VERSION}
+  REQUIRED COMPONENTS iostreams system
+)
+
+# Boost python handling - try and find both python and pythonXx (version suffixed).
+# Prioritize the version suffixed library, failing if neither exist.
 
 FIND_PACKAGE ( Boost ${MINIMUM_BOOST_VERSION}
-  REQUIRED COMPONENTS iostreams system ${BOOST_PYTHON_LIB}
+  QUIET COMPONENTS python${Python_VERSION_MAJOR}${Python_VERSION_MINOR}
 )
 
+IF ( TARGET Boost::python${Python_VERSION_MAJOR}${Python_VERSION_MINOR} )
+  SET ( BOOST_PYTHON_LIB "python${Python_VERSION_MAJOR}${Python_VERSION_MINOR}" )
+  MESSAGE ( STATUS "Found boost_python${Python_VERSION_MAJOR}${Python_VERSION_MINOR}" )
+ELSE ()
+  FIND_PACKAGE ( Boost ${MINIMUM_BOOST_VERSION}
+    QUIET COMPONENTS python
+  )
+  IF ( TARGET Boost::python )
+    SET ( BOOST_PYTHON_LIB "python" )
+    MESSAGE ( STATUS "Found non-suffixed boost_python, assuming to be python version "
+      "\"${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}\" compatible" )
+  ELSE ()
+    MESSAGE ( FATAL_ERROR "Unable to find boost_python or "
+      "boost_python${Python_VERSION_MAJOR}${Python_VERSION_MINOR}." )
+  ENDIF ()
+ENDIF ()
+
 IF ( UNIX )
   FIND_PACKAGE ( Threads REQUIRED )
 ENDIF ()